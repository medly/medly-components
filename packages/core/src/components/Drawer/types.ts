import { HTMLProps, WithStyle } from '@medly-components/utils';
import { ScrollActionTypes } from '../Modal/scrollStateReducer/types';
import { DrawerFooterProps } from './Footer/types';

export interface DrawerProps extends HTMLProps<HTMLDivElement> {
    open?: boolean;
    position?: 'left' | 'right';
    onClose: () => void;
    width?: string;
    withOverlay?: boolean;
}

export interface DrawerStaticProps {
    Content: React.FC & WithStyle;
    Footer: React.FC<DrawerFooterProps> & WithStyle;
    Header: React.FC & WithStyle;
}

interface ScrollState {
    scrolledToTop: boolean;
    scrolledToBottom: boolean;
    scrollPosition: number;
}

export interface DrawerContextType {
    id: string;
    scrollState: ScrollState;
    dispatch: React.Dispatch<ScrollActionTypes>;
    onClose: () => void;
}

export type DrawerStyledProps = { position: string; width: string; open?: boolean };
