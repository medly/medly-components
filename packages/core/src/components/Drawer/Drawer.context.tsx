/* eslint-disable @typescript-eslint/no-empty-function */
import { createContext } from 'react';
import { DrawerContextType } from './types';

export const defaultValues: DrawerContextType = {
    id: 'default-id',
    scrollState: {
        scrolledToTop: true,
        scrolledToBottom: false,
        scrollPosition: 0
    },
    dispatch: (): void => {},
    onClose: (): void => {}
};

export const DrawerContext = createContext<DrawerContextType>(defaultValues);
