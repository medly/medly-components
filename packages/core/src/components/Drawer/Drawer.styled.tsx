import { styled } from '@medly-components/utils';
import { DrawerStyledProps } from './types';

export const DrawerBackground = styled('div')<{ open?: boolean; withOverlay?: boolean }>`
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    z-index: 10;
    position: fixed;
    animation: ${({ withOverlay, open }) =>
        withOverlay ? `${open ? 'drawerFadeIn' : 'drawerFadeOut'} 0.4s cubic-bezier(0, 0, 0.33, 1) forwards` : 'none'};

    @keyframes drawerFadeIn {
        0% {
            background-color: transparent;
        }
        100% {
            background-color: ${({ theme }) => theme.drawer.overlayColor};
        }
    }

    @keyframes drawerFadeOut {
        0% {
            background-color: ${({ theme }) => theme.drawer.overlayColor};
        }
        100% {
            background-color: transparent;
        }
    }
`;

export const DrawerStyled = styled('aside')<DrawerStyledProps>`
    display: flex;
    flex-direction: column;
    height: 100vh;
    background-color: ${({ theme }) => theme.colors.white};
    width: ${props => props.width};
    position: fixed;
    top: 0;
    box-shadow: ${({ open }) => open && '0 4px 32px 0 rgb(67, 84, 101, 50%)'};
    box-sizing: border-box;
    animation: ${({ open }) => `${open ? 'drawerSlideIn' : 'drawerSlideOut'} 0.4s cubic-bezier(0, 0, 0.33, 1) forwards`};

    @keyframes drawerSlideIn {
        0% {
            left: ${({ width, position }) => position === 'left' && `-${width}`};
            right: ${({ width, position }) => position === 'right' && `-${width}`};
        }
        100% {
            left: ${({ position }) => position === 'left' && 0};
            right: ${({ position }) => position === 'right' && 0};
        }
    }

    @keyframes drawerSlideOut {
        0% {
            left: ${({ position }) => position === 'left' && 0};
            right: ${({ position }) => position === 'right' && 0};
        }
        100% {
            left: ${({ width, position }) => position === 'left' && `-${width}`};
            right: ${({ width, position }) => position === 'right' && `-${width}`};
        }
    }
`;
