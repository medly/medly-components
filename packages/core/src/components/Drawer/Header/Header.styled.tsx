import { ClearIcon } from '@medly-components/icons';
import { rgba } from 'polished';
import styled, { css } from 'styled-components';

export const HeaderStyled = styled('div')<{ showShadow: boolean }>`
    padding: 1.9rem 2.4rem 2.2rem;
    position: relative;
    box-shadow: ${({ theme, showShadow }) => showShadow && `0 1.8rem 1.6rem -1.6rem ${rgba(theme.colors.grey[400], 0.6)}`};
`;

const getStyle = ({ styleType }: { styleType: 'default' | 'hovered' | 'pressed' }) => css`
    background-color: ${({ theme }) => theme.drawer.closeIcon.bgColor[styleType]};
    * {
        fill: ${({ theme }) => theme.drawer.closeIcon.color[styleType]};
    }
`;

export const CloseIconStyled = styled(ClearIcon)`
    position: absolute;
    right: 1.4rem;
    top: 1.5rem;
    border-radius: ${({ theme }) => theme.drawer.closeIcon.borderRadius};

    ${props => getStyle({ ...props, styleType: 'default' })}

    &:hover {
        ${props => getStyle({ ...props, styleType: 'hovered' })}
        box-shadow:  0 2px 8px ${({ theme }) => rgba(theme.drawer.closeIcon.bgColor.hovered, 0.5)}
    }

    &:active {
        ${props => getStyle({ ...props, styleType: 'pressed' })}
        box-shadow: none;
    }
`;
