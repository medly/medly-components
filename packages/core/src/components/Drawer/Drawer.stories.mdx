import { Drawer } from './Drawer';
import { Preview, Story, Meta, Props } from '@storybook/addon-docs/blocks';
import * as stories from './Drawer.stories';

<Meta title="Core" component={Drawer} parameters={{ jest: ['Drawer.test.tsx'] }} />

# Drawer

A `Drawer` component is a modal anchored to the left or right edge of the screen. It consists of a Header, Content, and Footer. 

```typescript
<Drawer open={drawerState} onClose={changeDrawerState}>
    <Drawer.Header>Add Filters</Drawer.Header>
    <Drawer.Content>CONTENT</Drawer.Content>
    <Drawer.Footer>
        <Button variant="flat" onClick={onClearAll}>
            Clear All
        </Button>
    </Drawer.Footer>
</Drawer>
```

Shadow appears on the Header and Actions section if you scroll through the content.

## Example

Use the example given below: 

<Preview withToolbar>
    <Story name="Drawer">{stories.Basic()}</Story>
</Preview>

> Note: To view the code click [here](https://github.com/medly/medly-components/blob/master/packages/core/src/components/Drawer/Drawer.stories.tsx)

## Props

This component offers the following props mentioned in the table below:

<Props of={Drawer} />

## Drawer Footer Props

The drawer footer prop is given in the table below:

<Props of={Drawer.Footer} />
