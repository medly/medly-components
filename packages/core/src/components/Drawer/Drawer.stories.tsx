import { defaultTheme } from '@medly-components/theme/src';
import { styled } from '@medly-components/utils';
import { boolean, select, text } from '@storybook/addon-knobs';
import React, { useCallback, useState } from 'react';
import Button from '../Button';
import CheckboxGroup from '../CheckboxGroup';
import DateRangePicker from '../DateRangePicker';
import MultiSelect from '../MultiSelect';
import Text from '../Text';
import { Drawer } from './Drawer';

const FilterHeaderStyled = styled('div')`
    display: flex;
    justify-content: space-between;
    align-items: center;
    margin: 2.4rem 0 1.5rem;

    ${Button.Style} {
        padding: 0;
    }
`;

const FilterHeader: React.FC<{ label: string; onClear: () => void; showClear: boolean }> = React.memo(({ label, onClear, showClear }) => (
    <FilterHeaderStyled>
        <Text textVariant="h5" textColor={defaultTheme.colors.grey[600]}>
            {label}
        </Text>
        {showClear && (
            <Button variant="flat" onClick={onClear}>
                <Text textVariant="button2" textColor={defaultTheme.colors.blue[500]}>
                    Clear
                </Text>
            </Button>
        )}
    </FilterHeaderStyled>
));
FilterHeader.displayName = 'FilterHeader';

export const Basic = () => {
    const [drawerState, setDrawerState] = useState(false),
        [contract, setContract] = useState({ startDate: null, endDate: null }),
        [locations, setLocation] = useState(['Delhi']),
        [brands, setBrands] = useState(['Jaguar']),
        locationOptions = [
            { value: 'Delhi', label: 'Delhi' },
            { value: 'Pune', label: 'Pune' }
        ],
        brandOptions = [
            { value: 'Jaguar', label: 'Jaguar' },
            { value: 'Audi', label: 'Audi' },
            { value: 'Ferrari', label: 'Ferrari' },
            { value: 'GMC', label: 'GMC' },
            { value: 'Ford', label: 'Ford' },
            { value: 'Kia', label: 'Kia' },
            { value: 'Jeep', label: 'Jeep' },
            { value: 'Citroen', label: 'Citroen' },
            { value: 'BMW', label: 'BMW' },
            { value: 'Land Rover', label: 'Land Rover' },
            { value: 'Maserati', label: 'Maserati' },
            { value: 'Pagani', label: 'Pagani' },
            { value: 'Tesla', label: 'Tesla' },
            { value: 'Porsche', label: 'Porsche' },
            { value: 'Lamborghini', label: 'Lamborghini' },
            { value: 'Chevrolet', label: 'Chevrolet' },
            { value: 'Bentley', label: 'Bentley' },
            { value: 'McLaren', label: 'McLaren' },
            { value: 'Honda', label: 'Honda' },
            { value: 'Hyundai', label: 'Hyundai' },
            { value: 'Fiat', label: 'Fiat' }
        ];

    const clearDates = useCallback(() => setContract({ startDate: null, endDate: null }), []),
        clearLocation = useCallback(() => setLocation([]), []),
        clearBrands = useCallback(() => setBrands([]), []),
        clearAll = useCallback(() => {
            clearDates();
            clearLocation();
            clearBrands();
        }, [setLocation, setContract, setBrands]),
        showDrawer = useCallback(() => setDrawerState(true), []),
        hideDrawer = useCallback(() => setDrawerState(false), []);

    return (
        <>
            <Button onClick={showDrawer}>Click to Open</Button>
            <Drawer
                open={drawerState}
                onClose={hideDrawer}
                width={text('Width', '40rem')}
                position={select('Position', ['left', 'right'], 'right')}
                withOverlay={boolean('With Overlay', true)}
            >
                <Drawer.Header>Add Filters</Drawer.Header>
                <Drawer.Content>
                    <FilterHeader label="CONTRACT" onClear={clearDates} showClear={!!contract.startDate || !!contract.endDate} />
                    <DateRangePicker value={contract} onChange={setContract} fullWidth withSingleMonth />
                    <FilterHeader label="LOCATION" onClear={clearLocation} showClear={locations.length > 0} />
                    <MultiSelect options={locationOptions} values={locations} onChange={setLocation} fullWidth />
                    <FilterHeader label="BRAND" onClear={clearBrands} showClear={brands.length > 0} />
                    <CheckboxGroup fullWidth options={brandOptions} values={brands} onChange={setBrands} />
                </Drawer.Content>
                <Drawer.Footer alignItems="center">
                    <Button variant="flat" onClick={clearAll}>
                        Clear All
                    </Button>
                </Drawer.Footer>
            </Drawer>
        </>
    );
};
