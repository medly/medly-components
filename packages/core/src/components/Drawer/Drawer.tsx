import { useKeyPress, WithStyle } from '@medly-components/utils';
import React, { useCallback, useEffect, useReducer, useState } from 'react';
import { reducer } from '../Modal/scrollStateReducer';
import Content from './Content';
import { DrawerContext } from './Drawer.context';
import { DrawerBackground, DrawerStyled } from './Drawer.styled';
import Footer from './Footer';
import Header from './Header';
import { DrawerProps, DrawerStaticProps } from './types';

const Component: React.FC<DrawerProps> = React.memo(
    React.forwardRef(({ id, onClose, open, width, children, withOverlay, position, ...props }, ref) => {
        const isEscPressed = useKeyPress('Escape'),
            [shouldRender, setRenderState] = useState(open),
            [scrollState, dispatch] = useReducer(reducer, { scrolledToTop: true, scrolledToBottom: false, scrollPosition: 0 });

        const stopPropagation = useCallback((e: React.MouseEvent) => e.stopPropagation(), []),
            handleAnimationEnd = useCallback(() => !open && setRenderState(false), [open]);

        useEffect(() => {
            open && isEscPressed && onClose();
        }, [open, isEscPressed]);

        useEffect(() => {
            open && setRenderState(true);
        }, [open]);

        return shouldRender ? (
            <DrawerBackground onClick={onClose} ref={ref} open={open} {...props} id={`${id}-overlay`} withOverlay={withOverlay}>
                <DrawerStyled
                    id={id!}
                    position={position!}
                    onClick={stopPropagation}
                    width={width!}
                    open={open}
                    onAnimationEnd={handleAnimationEnd}
                >
                    <DrawerContext.Provider value={{ id: id!, scrollState, dispatch, onClose }}>{children}</DrawerContext.Provider>
                </DrawerStyled>
            </DrawerBackground>
        ) : null;
    })
);

Component.displayName = 'Drawer';
Component.defaultProps = {
    id: 'medly-drawer',
    open: false,
    position: 'right',
    width: '40rem',
    withOverlay: true
};
export const Drawer: React.FC<DrawerProps> & WithStyle & DrawerStaticProps = Object.assign(Component, {
    Style: DrawerBackground,
    Content,
    Footer,
    Header
});
