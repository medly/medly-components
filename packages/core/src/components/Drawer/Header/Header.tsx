import { isValidStringOrNumber, WithStyle } from '@medly-components/utils';
import React, { useContext } from 'react';
import Text from '../../Text';
import { DrawerContext } from '../Drawer.context';
import { CloseIconStyled, HeaderStyled } from './Header.styled';

const Component: React.FC = React.memo(({ children, ...restProps }) => {
    const { id, scrollState, onClose } = useContext(DrawerContext);
    return (
        <HeaderStyled showShadow={!scrollState.scrolledToTop} {...restProps} id={`${id}-header`}>
            {React.Children.map(children, c => {
                return isValidStringOrNumber(c) ? <Text textVariant="h3">{c}</Text> : c;
            })}
            <CloseIconStyled onClick={onClose} size="M" variant="solid" id={`${id}-close-icon`} title={`${id}-close-icon`} />
        </HeaderStyled>
    );
});

Component.displayName = 'DrawerHeader';
export const Header: React.FC & WithStyle = Object.assign(Component, { Style: HeaderStyled });
