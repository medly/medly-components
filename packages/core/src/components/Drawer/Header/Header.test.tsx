import { render } from '@test-utils';
import { DrawerContext } from '../Drawer.context';
import { Header } from './Header';

describe('Drawer Header Test', () => {
    const contextValues = {
        id: 'medly-drawer',
        scrollState: {
            scrolledToTop: true,
            scrolledToBottom: false,
            scrollPosition: 0
        },
        dispatch: jest.fn(),
        onClose: jest.fn()
    };

    it('should render text if string is passed', () => {
        const { container } = render(
            <DrawerContext.Provider value={contextValues}>
                <Header>Demo Header</Header>
            </DrawerContext.Provider>
        );
        expect(container).toMatchSnapshot();
    });

    it('should render component if its passed', () => {
        const { container } = render(
            <DrawerContext.Provider value={contextValues}>
                <Header>
                    <div>Render DIV</div>
                </Header>
            </DrawerContext.Provider>
        );
        expect(container).toMatchSnapshot();
    });
});
