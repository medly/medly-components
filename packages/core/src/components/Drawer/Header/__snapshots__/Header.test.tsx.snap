// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Drawer Header Test should render component if its passed 1`] = `
.c1 {
  overflow: visible;
  font-size: 2.4rem;
  -webkit-transition: all 100ms linear;
  transition: all 100ms linear;
  cursor: pointer;
  padding: 0.8rem;
  border-radius: 50%;
  background-color: #E7F0FF;
}

.c1 * {
  fill-opacity: 1;
  -webkit-transition: all 100ms linear;
  transition: all 100ms linear;
  fill: #607890;
}

.c0 {
  padding: 1.9rem 2.4rem 2.2rem;
  position: relative;
}

.c2 {
  position: absolute;
  right: 1.4rem;
  top: 1.5rem;
  border-radius: 50%;
  background-color: #ffffff;
}

.c2 * {
  fill: #607890;
}

.c2:hover {
  background-color: #eff2f4;
  box-shadow: 0 2px 8px rgba(239,242,244,0.5);
}

.c2:hover * {
  fill: #607890;
}

.c2:active {
  background-color: #dfe4e9;
  box-shadow: none;
}

.c2:active * {
  fill: #435465;
}

<div>
  <div
    class="c0"
    id="medly-drawer-header"
  >
    <div>
      Render DIV
    </div>
    <svg
      class="c1 c2"
      fill="none"
      height="1em"
      id="medly-drawer-close-icon"
      title="medly-drawer-close-icon"
      variant="solid"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M18.3 5.71a.996.996 0 0 0-1.41 0L12 10.59 7.11 5.7A.996.996 0 1 0 5.7 7.11L10.59 12 5.7 16.89a.996.996 0 1 0 1.41 1.41L12 13.41l4.89 4.89a.996.996 0 1 0 1.41-1.41L13.41 12l4.89-4.89c.38-.38.38-1.02 0-1.4z"
        fill="#000"
        fill-opacity=".54"
      />
    </svg>
  </div>
</div>
`;

exports[`Drawer Header Test should render text if string is passed 1`] = `
.c1 {
  margin: 0;
  color: inherit;
  font-size: 2.4rem;
  font-weight: 400;
  -webkit-letter-spacing: -0.04rem;
  -moz-letter-spacing: -0.04rem;
  -ms-letter-spacing: -0.04rem;
  letter-spacing: -0.04rem;
  line-height: 3.2rem;
  text-align: initial;
}

.c2 {
  overflow: visible;
  font-size: 2.4rem;
  -webkit-transition: all 100ms linear;
  transition: all 100ms linear;
  cursor: pointer;
  padding: 0.8rem;
  border-radius: 50%;
  background-color: #E7F0FF;
}

.c2 * {
  fill-opacity: 1;
  -webkit-transition: all 100ms linear;
  transition: all 100ms linear;
  fill: #607890;
}

.c0 {
  padding: 1.9rem 2.4rem 2.2rem;
  position: relative;
}

.c3 {
  position: absolute;
  right: 1.4rem;
  top: 1.5rem;
  border-radius: 50%;
  background-color: #ffffff;
}

.c3 * {
  fill: #607890;
}

.c3:hover {
  background-color: #eff2f4;
  box-shadow: 0 2px 8px rgba(239,242,244,0.5);
}

.c3:hover * {
  fill: #607890;
}

.c3:active {
  background-color: #dfe4e9;
  box-shadow: none;
}

.c3:active * {
  fill: #435465;
}

<div>
  <div
    class="c0"
    id="medly-drawer-header"
  >
    <h3
      class="c1"
    >
      Demo Header
    </h3>
    <svg
      class="c2 c3"
      fill="none"
      height="1em"
      id="medly-drawer-close-icon"
      title="medly-drawer-close-icon"
      variant="solid"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M18.3 5.71a.996.996 0 0 0-1.41 0L12 10.59 7.11 5.7A.996.996 0 1 0 5.7 7.11L10.59 12 5.7 16.89a.996.996 0 1 0 1.41 1.41L12 13.41l4.89 4.89a.996.996 0 1 0 1.41-1.41L13.41 12l4.89-4.89c.38-.38.38-1.02 0-1.4z"
        fill="#000"
        fill-opacity=".54"
      />
    </svg>
  </div>
</div>
`;
